### ISSUE TITLE
calculation of labelnumberwidth does not respect format of labelnumber

### ISSUE THREAD

**Issue by moewew** [plk/biblatex#342]
In [Use roman numerals for cite in BibLaTeX](http://tex.stackexchange.com/q/271527/35864) I noticed that the code used to calculate the `labelnumberwidth` does not respect the format of `labelnumber`, it seems to be taken bare, regardless of any `\DeclareFieldFormat{labelnumber}`.

The effect can be seen in the MWE
```
\documentclass{article}
\usepackage[style=numeric,backend=biber]{biblatex}

\DeclareFieldFormat{labelnumber}{\RN{#1}}

\addbibresource{biblatex-examples.bib}
\begin{document}
\cite{sigfridsson,geer,wilde,worman,knuth:ct:a}
\printbibliography
\end{document}
```

This can be solved by adding the directive `\csuse{abx@ffd@*@labelnumber}` to [line 5620 of `biblatex2.sty`](https://github.com/plk/biblatex/blob/master/tex/latex/biblatex/biblatex2.sty#L5620), such that the total definition of `\blx@bbl@labelnumber` now reads

```
\def\blx@bbl@labelnumber{%
  % Used at endentry, when reading .bbl
  % only do this for the global sorting list otherwise we increment
  % the labelnumbers for each sortlist
  \ifdefempty\abx@field@localnumber
  {}% only if omitnumbers=true
  {\ifundef\abx@field@shorthand
    {\iftoggle{blx@defernumbers}
      % only if defernumbers=true, we have to define localnumber to
      % something to stop labelnumberwidth def complaining on first
      % post-backend run since there will be nothing in the .aux yet
      % for defernums localnumbers. On the second post-backend run,
      % it will have a value correct for generating labelnumberwidth
      % We also have to define a dummy value for labelnumber here for
      % first post-backend runs otherwise references to labelnumber
      % in cite styles like numeric-comp will die.
      {\ifundef\abx@field@localnumber
        {\numdef\abx@field@localnumber{0}%
          \blx@bbl@fielddef{labelnumber}{0}}
        {}}
      {\csnumgdef{blx@labelnumber@\the\c@refsection @\blx@slist@scheme}{%
          \csuse{blx@labelnumber@\the\c@refsection @\blx@slist@scheme}+1}%
        \edef\abx@field@localnumber{%
          \csuse{blx@labelnumber@\the\c@refsection @\blx@slist@scheme}}%
        \blx@bbl@fieldedef{labelnumber}{\abx@field@localnumber}}%
      \iftoggle{blx@skipbib}
        {}
        {\blx@setlabwidth{\labelnumberwidth}{%
           \csuse{abx@ffd@*@labelnumberwidth}{%
              \ifdef\abx@field@prefixnumber
                {\csuse{abx@ffd@*@prefixnumber}{\abx@field@prefixnumber}}
                {}%
              \csuse{abx@ffd@*@labelnumber}{\abx@field@localnumber}}}}}
    {\csgappto\blx@bbl@data{\let\abx@field@labelnumber\abx@field@shorthand}%
      \iftoggle{blx@skipbib}
        {}
        {\blx@setlabwidth{\labelnumberwidth}{%
           \csuse{abx@ffd@*@labelnumberwidth}{\abx@field@shorthand}}}}}}
```

**Comment by moewew** [comment]
I also noted that the label width calculation does not take into account type-specific formats, it will always use the format declared for `*`.

The following allows for that as well (I'm sure the coding style does not quite fit)

```
\def\blx@bbl@labelnumber{%
  % Used at endentry, when reading .bbl
  % only do this for the global sorting list otherwise we increment
  % the labelnumbers for each sortlist
  \ifdefempty\abx@field@localnumber
  {}% only if omitnumbers=true
  {%
   \ifcsundef{abx@ffd@\blx@imc@thefield{entrytype}@labelnumberwidth}
     {\letcs{\@lnw@labelnumberwidth}{abx@ffd@*@labelnumberwidth}}
     {\letcs{\@lnw@labelnumberwidth}{abx@ffd@\blx@imc@thefield{entrytype}@labelnumberwidth}}
   \ifcsundef{abx@ffd@\blx@imc@thefield{entrytype}@prefixnumber}
     {\letcs{\@lnw@prefixnumber}{abx@ffd@*@prefixnumber}}
     {\letcs{\@lnw@prefixnumber}{abx@ffd@\blx@imc@thefield{entrytype}@prefixnumber}}
   \ifcsundef{abx@ffd@\blx@imc@thefield{entrytype}@labelnumber}
     {\letcs{\@lnw@labelnumber}{abx@ffd@*@labelnumber}}
     {\letcs{\@lnw@labelnumber}{abx@ffd@\blx@imc@thefield{entrytype}@labelnumber}}  
   \ifundef\abx@field@shorthand
    {\iftoggle{blx@defernumbers}
      % only if defernumbers=true, we have to define localnumber to
      % something to stop labelnumberwidth def complaining on first
      % post-backend run since there will be nothing in the .aux yet
      % for defernums localnumbers. On the second post-backend run,
      % it will have a value correct for generating labelnumberwidth
      % We also have to define a dummy value for labelnumber here for
      % first post-backend runs otherwise references to labelnumber
      % in cite styles like numeric-comp will die.
      {\ifundef\abx@field@localnumber
        {\numdef\abx@field@localnumber{0}%
          \blx@bbl@fielddef{labelnumber}{0}}
        {}}
      {\csnumgdef{blx@labelnumber@\the\c@refsection @\blx@slist@scheme}{%
          \csuse{blx@labelnumber@\the\c@refsection @\blx@slist@scheme}+1}%
        \edef\abx@field@localnumber{%
          \csuse{blx@labelnumber@\the\c@refsection @\blx@slist@scheme}}%
        \blx@bbl@fieldedef{labelnumber}{\abx@field@localnumber}}%
      \iftoggle{blx@skipbib}
        {}
        {\blx@setlabwidth{\labelnumberwidth}{%
           \csuse{@lnw@labelnumberwidth}{%
              \ifdef\abx@field@prefixnumber
                {\csuse{@lnw@prefixnumber}{\abx@field@prefixnumber}}
                {}%
              \csuse{@lnw@labelnumber}{\abx@field@localnumber}}}}}
    {\csgappto\blx@bbl@data{\let\abx@field@labelnumber\abx@field@shorthand}%
      \iftoggle{blx@skipbib}
        {}
        {\blx@setlabwidth{\labelnumberwidth}{%
           \csuse{@lnw@labelnumber}{\abx@field@shorthand}}}}}}
```

**Comment by plk** [comment]
Is there a reason why you wrapped `\abx@field@localnumber` in extra braces?

**Comment by moewew** [comment]
That would only be my limited knowledge of what is best practice (or even strictly speaking correct) here. I followed the example of `\csuse{abx@ffd@*@prefixnumber}{\abx@field@prefixnumber}` or are what braces are you talking about?
As I said, my coding style is more focused on what seems to work and I don't know anything about proper conventions, unfortunately.

**Comment by plk** [comment]
That's ok - have you tested this much with defernumbers etc?

**Comment by moewew** [comment]
I have not done extensive testing, yet. But I just discovered that in my second suggestion the last line should have read

```
\csuse{@lnw@labelnumberwidth}{\abx@field@shorthand}}}}}}
```

All other ad-hoc test I ran looked OK.

**Comment by plk** [comment]
Ok, thanks - perhaps you can do a bit of testing with the defernums option and then I'm happy to use this.

**Comment by moewew** [comment]
I checked the examples `32-style-numeric-verb.tex`, `16-numeric-prefixed-1.tex` and `17-numeric-prefixed-2.tex` and could see no difference at all. I also checked some random MWEs using `defernumbers=true` I found on TeX.SX, all of them were fine as well.

I think though that if the code above is buggy it will show regardless of the `defernumbers` setting, because only the `\blx@setlabwidth{\labelnumberwidth}` part was effectively changed - it should not tinker with the actual label generation.
